src/AbstractGenerator.h

```diff
@@ -0,0 +1,190 @@
+#include "Pipeline.h"
+#include "Schedule.h"
+#include "Target.h"
+#include "Type.h"
+
+namespace Halide {
+
+class GeneratorContext;
+
+namespace Internal {
+
+enum class ArgInfoKind { Scalar,
+                         Function,
+                         Buffer };
+
+enum class ArgInfoDirection { Input,
+                              Output };
+
+using ExternsMap = std::map<std::string, ExternalCode>;
+
+/**
+ * AbstractGenerator is an ABC that defines the API a Generator must provide
+ * to work with the existing Generator infrastructure (GenGen, RunGen,
+ * Generator Stubs). The existing Generator<>-based instances all implement
+ * this API, but any other code that implements this (and uses RegisterGenerator
+ * to register itself) should be indistinguishable from a user perspective.
+ *
+ * An AbstractGenerator is meant to be "single-use"; typically lifetimes will be
+ * something like:
+ * - create an instance (with a specific Target)
+ * - optionally set GeneratorParam values
+ * - optionally re-bind inputs (if using in JIT or Stub modes)
+ * - call build_pipeline()
+ * - optionally call get_output_func() to get the output(s) (if using in JIT or Stub modes)
+ * - discard the instance
+ *
+ * AbstractGenerators should be fairly cheap to instantiate! Don't try to re-use
+ * one by re-setting inputs and calling build_pipeline() multiple times.
+ *
+ * Note that an AbstractGenerator instance is (generally) stateful in terms of the order
+ * that methods should be called; calling the methods out of order may cause
+ * assert-fails or other undesirable behavior. Read the method notes carefully!
+ */
+class AbstractGenerator {
+public:
+    virtual ~AbstractGenerator() = default;
+
+    /** ArgInfo is a struct to contain name-and-type information for the inputs and outputs to
+     * the Pipeline that build_pipeline() will return.
+     *
+     * Note that this looks rather similar to Halide::Argument, but unfortunately
+     * that is not a good fit here, as it cannot represent Func inputs (only
+     * Buffer and Scalar), nor can it really handle Outputs.
+     *
+     * TODO: name is suboptimal. Is there a better one?
+     */
+    struct ArgInfo {
+        std::string name;
+        ArgInfoDirection dir = ArgInfoDirection::Input;
+        ArgInfoKind kind = ArgInfoKind::Scalar;
+        // Note that this can have multiple entries for Tuple-valued Inputs or Outputs
+        std::vector<Type> types;
+        int dimensions = 0;
+    };
+
+    /** Return the name of this Generator. (This should always be the name
+     * used to register it.) */
+    virtual std::string get_name() = 0;
+
+    /** Return the Target, autoscheduler flag, and MachineParams that this Generator
+     * was created with. Always legal to call on any AbstractGenerator instance,
+     * regardless of what other methods have been called. (All AbstractGenerator instances
+     * are expected to be created with immutable values for these, which can't be
```

Comment:
The immutable bit troubles me. There should be some mechanism to set or modify these in the Generator itself (or in the registration of the generator?) for those of us build systems incapable of supporting ad-hoc per-generator configuration. Currently this can be done in the configure() method.

Reply:
I think this comment will need updating depending on how #6760 resolves.

Reply:
Agreed.

Reply:
I've updated the comments here to reflect the landing of #6760.